### PR TITLE
Centralize WTS refresh token initialization

### PR DESCRIPTION
**Pull request by pieterlukasse**
Jira Ticket: [VADC-154](https://ctds-planx.atlassian.net/browse/VADC-154)

### Improvements
- WTS is a useful service for other backend components that need to access gen3 services using a token bound to a specific user account.

### Dependency updates
- WTS becomes a dependency for data-portal in general, instead of just for the /workspace subsection of the portal.
  - The change IS backwards compatible, since this is controlled by a new optional "feature flag" called `workspaceTokenServiceRefreshTokenAtLogin`. If this is set to `true` it will refresh the WTS token directly at portal login (recommended mode). If not set, this refresh happens only when the user enters the workspace section of the portal (default/old/previous mode).

### Deployment changes

- this change assumes WTS is already a core/central service. The dependency is hereby moved to the homepage (instead of the ./workspace page)

### CODE REVIEW

**Comment by pieterlukasse** [comment]
@m0nhawk @kmhernan who else needs to be added as reviewer here?

**Comment by Avantol13** [comment]
I'm not sure I agree with the concept here. WTS should not be a hard dependency of portal. We have data commons that need portal but don't need WTS

**Comment by Avantol13** [comment]
The assumption that WTS is configured and used everywhere is not a safe assumption

**Comment by pieterlukasse** [comment]
@Avantol13 thanks for the review! You bring up a good point, and I thought about this as well when making this PR. One reason I added it here is because I see other workspace related items in `src/index.jsx` already. Do these then also not belong there, or is there a part of "workspace" that functions and is useful without WTS?

**Comment by Avantol13** [comment]
> @Avantol13 thanks for the review! You bring up a good point, and I thought about this as well when making this PR. One reason I added it here is because I see other workspace related items in `src/index.jsx` already. Do these then also not belong there, or is there a part of "workspace" that functions and is useful without WTS?

Honestly I don't think workspace things should necessarily be there either, but adding more coupling (for example to WTS) is just making things worse in my opinion. I don't have a ton of portal experience so ultimately I need to defer to @mfshao 

imo portal should support commons that don't have workspaces at all, commons that only have workspaces for themselves (e.g. no need to WTS), and it should support ecosystems where we want workspaces and WTS. 

We have running, production examples of all of those things, and it seems like this PR presumes that we want everything to be an ecosystem with workspaces and WTS, and that just isn't the case.

**Comment by pieterlukasse** [comment]
thanks @Avantol13, that makes sense. I agree that in this case probably other references to workspaces should also be removed. @mfshao what do you think?

**Comment by pieterlukasse** [comment]
@mfshao thanks for your comment. I added this commit https://github.com/uc-cdis/data-portal/pull/1095/commits/14562674d3fb5d43d75d662d97cd73ab9abce9d6 which will trigger the initialization/refresh based on a very specific feature flag (I called it `workspaceTokenServiceRefreshTokenAtLogin`). This is still based on @Avantol13 's comment that some installations can have workspace without needing WTS. Please let me know what you think.

File: docs/portal_config.md
Comment (Avantol13) [inline docs/portal_config.md:185]: I feel like the default here should be `false` if we're trying to be backwards compatible w/o needing to update config. That way, people that want to take advantage of this can turn it on in there config but there's no change to those who want the old previous mode

Comment (pieterlukasse) [reply docs/portal_config.md:185]: Yes, the default is "not set", so `isEnabled('workspaceTokenServiceRefreshTokenAtLogin')` should return `false`.

File: src/Workspace/index.jsx
Comment (mfshao) [inline src/Workspace/index.jsx:77]: so, if this new feature is enabled, when user navigates from a page to workspace, another wts call will be dispatched? seems like a bit redundant

File: src/index.jsx
Comment (mfshao) [inline src/index.jsx:72]: other index pages (which are being used by other portal bundles) also needs this change?

**Review by mfshao — COMMENTED** [review]
I don't think we support "workspace without WTS" a long while ago... 

But this PR has some linting issues that needs to be fixed

File: src/index.jsx
Comment (pieterlukasse) [reply src/index.jsx:72]: not sure. Is this one not used by all?

File: src/Workspace/index.jsx
Comment (pieterlukasse) [reply src/Workspace/index.jsx:77]: Good point... But I think that if we want to do this with a feature flag that is `false` by default, then we will need to keep some of this redundancy for backwards compatibility purposes. Also, an extra refresh is not so bad...it is better than _no_ refresh (which happens if users use the portal but don't go into the workspace for a long period - a scenario we have in our project).

Comment (pieterlukasse) [reply src/Workspace/index.jsx:77]: Please note that this diff is technically not really a code/behaviour change: I basically moved the code on the left into a generic function that can be used in different places.

Comment (pieterlukasse) [reply src/Workspace/index.jsx:77]: Another option is to make a breaking change and make a feature flag for workspace as a whole, and make this flag `true` by default. This would simplify the code, but would require some special attention from teams that have a deployment that does not use workspace/ WTS.

Comment (paulineribeyre) [reply src/Workspace/index.jsx:77]: why not check if `workspaceTokenServiceRefreshTokenAtLogin` is enabled, and only make the call if it's not?

File: src/index.jsx
Comment (paulineribeyre) [reply src/index.jsx]: no, some Data Commons have their own landing page, so they have their own index page (`covid19Index.jsx`, `nctIndex.jsx`, etc)

File: src/Workspace/index.jsx
Comment (pieterlukasse) [reply src/Workspace/index.jsx:77]: That could work, but IMO it is an unnecessary optimisation. But if you prefer this, I can easily add it. Let me know.

File: src/index.jsx
Comment (pieterlukasse) [reply src/index.jsx]: so do we need to add this change there as well?

File: src/Login/ProtectedContent.jsx
Comment (pieterlukasse) [reply src/Login/ProtectedContent.jsx:77]: @paulineribeyre would that be a problem? I think it is a really light call.

Comment (paulineribeyre) [reply src/Login/ProtectedContent.jsx:77]: Yes i think that's a problem, from a quick search i see 26 `ProtectedContents` in `src/index.jsx`, we should not make so many API calls every time someone loads the page. Can we make a single call to WTS when we portal is loaded?

File: src/index.jsx
Comment (paulineribeyre) [reply src/index.jsx]: The logic needs to happen for every Commons, not just the ones that use the default landing page

File: src/Workspace/index.jsx
Comment (paulineribeyre) [reply src/Workspace/index.jsx:77]: it makes sense to me... depending on the flag, the call is made either globally, or when the workspace page loads, and not both. But we can defer to @mfshao who brought this up in the first place

File: src/Login/ProtectedContent.jsx
Comment (pieterlukasse) [reply src/Login/ProtectedContent.jsx:77]: Interesting, since the same could be said regarding the call to `checkLoginStatus` just a few lines up:

```js
.then(
            () => this.checkLoginStatus(store, this.state)
```
Maybe that is why `checkLoginStatus` has [this code ](https://github.com/uc-cdis/data-portal/pull/1095/files#diff-c8ea73aaef1661450270bc0f167f41abd12c3d3a78f9894a48460f3727aeb0baR123-R126) to avoid repeated calls? Should we do something similar here? Or should we try to find another location for our init code?
```js
  checkLoginStatus = (store, initialState) => {
    const newState = { ...initialState };
    const nowMs = Date.now();
    newState.authenticated = true;
    newState.redirectTo = null;
    newState.user = store.getState().user;
    if (nowMs - lastAuthMs < 60000) {
      // assume we're still logged in after 1 minute ...
      return Promise.resolve(newState);
    }
  ...
```
@mfshao what would you suggest for this?

File: src/index.jsx
Comment (pieterlukasse) [reply src/index.jsx]: Ok, I'll add that

File: src/Workspace/index.jsx
Comment (mfshao) [reply src/Workspace/index.jsx:77]: That is the point I bring this up: we should be able to add a check so this didn't get called repeatly

File: src/Login/ProtectedContent.jsx
Comment (mfshao) [reply src/Login/ProtectedContent.jsx:77]: you can do something similar like this

or change in your wts init code so that if a user is not logged in, don't force them to do an OIDC dance with WTS

Comment (paulineribeyre) [reply src/Login/ProtectedContent.jsx:77]: i am curious why you chose to add it to `ProtectedContent` instead of `index.jsx` like in your initial code?

Comment (pieterlukasse) [reply src/Login/ProtectedContent.jsx:77]: @paulineribeyre I was asked by @mfshao to do this...I now realise this was in slack and is not here in the PR comments. 

@mfshao do you mind replying to this here?

Comment (pieterlukasse) [reply src/Login/ProtectedContent.jsx:77]: @paulineribeyre @mfshao I added a proposal here in this last commit: https://github.com/uc-cdis/data-portal/pull/1095/commits/82b7b20dd878e31ad3137ba41619ed758e315f65. Please let me know what you think. It is based on the current solution for `checkLoginStatus` in `ProtectedContent.jsx`

File: src/Workspace/index.jsx
Comment (pieterlukasse) [reply src/Workspace/index.jsx]: Fixed here: https://github.com/uc-cdis/data-portal/pull/1095/commits/882043f31b3f99be0e18cc94337e64b8ba8ae442

File: src/index.jsx
Comment (pieterlukasse) [reply src/index.jsx]: Or not... This change has been reverted in favor of the one in `src/Login/ProtectedContent.jsx` 😅

**Review by paulineribeyre — COMMENTED** [review]
lgtm, deferring to @mfshao

File: src/Workspace/WorkspaceRefreshToken.js
Comment (mfshao) [inline src/Workspace/WorkspaceRefreshToken.js:19]: Notice that the redirect here only contains `window.location.pathname`, so if the page that the user is at contains some query params in its URL, they will be lost

This was not a problem before because WTS login redirect only happens in certian pages that don't need query params. But if you are moving this redirect so it can happen virtually from any portal pages, you need to accomodate accordingly, because this will have a product-wise impact

Also there are some location and history based redirect/state restore happens in other pages of portal. We should make sure this change doens't breaks them

**Review by mfshao — COMMENTED** [review]
I think we should test this features more. At least we should configure the portal to have some public/protected pages and try to login with user who already has WTS connected and who doesn't have WTS connected. To see if there will be any unexpected behaviors

**Comment by mfshao** [comment]
Also the tests need to get fixed before it can be merged

File: src/Workspace/WorkspaceRefreshToken.js
Comment (pieterlukasse) [reply src/Workspace/WorkspaceRefreshToken.js]: hi @mfshao, as discussed, I've update the redirect code to use the same logic as is used in the `src/Login/Login.jsx` page. See latest commit added to this PR: https://github.com/uc-cdis/data-portal/pull/1095/commits/8950eec53745de249fe0445cc6dffdce212fbfdf

Please let me know if this is the change you were looking for.

File: src/Workspace/index.jsx
Comment (mfshao) [inline src/Workspace/index.jsx:78]:
```suggestion
    // really needed, since it has already happened at login, so just call the callback:
```

**Review by mfshao — DISMISSED** [review]
looks cool

**Review by pieterlukasse — COMMENTED** [review]
tested locally:
- when `workspaceTokenServiceRefreshTokenAtLogin` is `false`: WTS refresh occurs only at `/workspace`
- when `workspaceTokenServiceRefreshTokenAtLogin` is `true`: WTS refresh occurs only at login / browser refresh event

 ✅

**Review by mfshao — APPROVED** [review]

**Pull request merged**
